<a href="https://colab.research.google.com/github/bhavanaramakrishna6/DL_ASSIGNMENT/blob/main/DL_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#SGD WITH FASHION MNIST

In [3]:
# Using SGD Optimizer on Fashion-MNIST

import numpy as np
import pandas as pd
import altair as alt
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.utils import to_categorical

# Load Fashion-MNIST dataset
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Normalize pixel values
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encode labels
y_train_cat = to_categorical(y_train, 10)
y_test_cat = to_categorical(y_test, 10)

# Build the model
model_sgd = Sequential()
model_sgd.add(Flatten(input_shape=(28, 28)))
model_sgd.add(Dense(128, activation='relu'))
model_sgd.add(Dense(10, activation='softmax'))

# Compile model using SGD optimizer
model_sgd.compile(
    loss='categorical_crossentropy',
    optimizer=SGD(learning_rate=0.01),
    metrics=['accuracy']
)

# Train the model
history_sgd = model_sgd.fit(
    X_train,
    y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1,
    verbose=1
)

# Extract loss values
errors_sgd = history_sgd.history['loss']

# Create DataFrame for visualization
df_sgd = pd.DataFrame({
    "errors": errors_sgd,
    "epoch": np.arange(1, len(errors_sgd) + 1)
})

# Visualization with Altair
alt.Chart(df_sgd).mark_line().encode(
    x="epoch",
    y="errors"
).properties(title="SGD Optimizer - Training Loss Curve (Fashion-MNIST)")
# Evaluate on test data
test_loss, test_accuracy = model_sgd.evaluate(X_test, y_test_cat, verbose=0)
print(f"SGD Optimizer Test Accuracy: {test_accuracy * 100:.2f}%")

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.6738 - loss: 1.0916 - val_accuracy: 0.7563 - val_loss: 0.7621
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7712 - loss: 0.7020 - val_accuracy: 0.7888 - val_loss: 0.6419
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7975 - loss: 0.6183 - val_accuracy: 0.8042 - val_loss: 0.5843
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8106 - loss: 0.5725 - val_accuracy: 0.8142 - val_loss: 0.5512
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8185 - loss: 0.5424 - val_accuracy: 0.8217 - val_loss: 0.5278
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8250 - loss: 0.5206 - val_accuracy: 0.8263 - val_loss: 0.5115
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8294 - loss: 0.5039 - val_accuracy: 0.8248 - val_loss: 0.4971
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8326 - loss: 0.4904 - val_accuracy: 0.

Test Accuracy: ~84%

Gradual improvement across epochs

Slight fluctuations but stable

Slower than advanced optimizers

Conclusion: Simple but not very fast

#BGD WITH FASHION MNIST

In [4]:
from tensorflow.keras.optimizers import SGD

model_bgd = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_bgd.compile(
    loss='categorical_crossentropy',
    optimizer=SGD(learning_rate=0.01),
    metrics=['accuracy']
)

history_bgd = model_bgd.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=len(X_train),   # FULL dataset
    validation_split=0.1,
    verbose=1
)

Epoch 1/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 3s 3s/step - accuracy: 0.0838 - loss: 2.3885 - val_accuracy: 0.0997 - val_loss: 2.3577
Epoch 2/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 343ms/step - accuracy: 0.0986 - loss: 2.3560 - val_accuracy: 0.1168 - val_loss: 2.3280
Epoch 3/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 288ms/step - accuracy: 0.1142 - loss: 2.3265 - val_accuracy: 0.1327 - val_loss: 2.3007
Epoch 4/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 237ms/step - accuracy: 0.1293 - loss: 2.2995 - val_accuracy: 0.1468 - val_loss: 2.2755
Epoch 5/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 243ms/step - accuracy: 0.1431 - loss: 2.2744 - val_accuracy: 0.1595 - val_loss: 2.2519
Epoch 6/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 248ms/step - accuracy: 0.1562 - loss: 2.2510 - val_accuracy: 0.1703 - val_loss: 2.2298
Epoch 7/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 237ms/step - accuracy: 0.1675 - loss: 2.2291 - val_accuracy: 0.1808 - val_loss: 2.2089
Epoch 8/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 245ms/step - accuracy: 0.1797 - loss: 2.2083 - val_accuracy: 0.1950 - val_loss: 2.

Accuracy improved very slowly (from ~8% → ~38%)

Only 1 update per epoch, so learning is inefficient

Loss decreases smoothly but very slowly

Worst performer

Conclusion: Not suitable for large datasets like Fashion-MNIST

#SGD WITH MOMENTUM

In [5]:
from tensorflow.keras.optimizers import SGD

model_momentum = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_momentum.compile(
    loss='categorical_crossentropy',
    optimizer=SGD(learning_rate=0.01, momentum=0.9),
    metrics=['accuracy']
)

history_momentum = model_momentum.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1
)

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.7694 - loss: 0.6717 - val_accuracy: 0.8202 - val_loss: 0.5091
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8387 - loss: 0.4631 - val_accuracy: 0.8415 - val_loss: 0.4399
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8509 - loss: 0.4265 - val_accuracy: 0.8573 - val_loss: 0.4058
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8589 - loss: 0.4026 - val_accuracy: 0.8575 - val_loss: 0.3985
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8634 - loss: 0.3849 - val_accuracy: 0.8575 - val_loss: 0.4053
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8704 - loss: 0.3687 - val_accuracy: 0.8653 - val_loss: 0.3774
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8734 - loss: 0.3564 - val_accuracy: 0.8670 - val_loss: 0.3665
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8774 - loss: 0.3441 - val_accuracy: 0.

Accuracy reached ~90%+ (train), ~88% val

Much faster convergence than SGD

Reduced oscillations in loss

Conclusion: Momentum improves speed and stability significantl

#MINI BATCH GD

In [6]:
# Using SGD Optimizer on Fashion-MNIST

import numpy as np
import pandas as pd
import altair as alt
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.utils import to_categorical

# Load Fashion-MNIST dataset
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Normalize pixel values
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encode labels
y_train_cat = to_categorical(y_train, 10)
y_test_cat = to_categorical(y_test, 10)

# Build the model
model_sgd = Sequential()
model_sgd.add(Flatten(input_shape=(28, 28)))
model_sgd.add(Dense(128, activation='relu'))
model_sgd.add(Dense(10, activation='softmax'))

# Compile model using SGD optimizer
model_sgd.compile(
    loss='categorical_crossentropy',
    optimizer=SGD(learning_rate=0.01),
    metrics=['accuracy']
)

# Train the model
history_sgd = model_sgd.fit(
    X_train,
    y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1,
    verbose=1
)

# Extract loss values
errors_sgd = history_sgd.history['loss']

# Create DataFrame for visualization
df_sgd = pd.DataFrame({
    "errors": errors_sgd,
    "epoch": np.arange(1, len(errors_sgd) + 1)
})

# Visualization with Altair
alt.Chart(df_sgd).mark_line().encode(
    x="epoch",
    y="errors"
).properties(title="SGD Optimizer - Training Loss Curve (Fashion-MNIST)")
# Evaluate on test data
test_loss, test_accuracy = model_sgd.evaluate(X_test, y_test_cat, verbose=0)
print(f"SGD Optimizer Test Accuracy: {test_accuracy * 100:.2f}%")

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.6500 - loss: 1.1193 - val_accuracy: 0.7510 - val_loss: 0.7724
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7664 - loss: 0.7131 - val_accuracy: 0.7857 - val_loss: 0.6444
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7943 - loss: 0.6238 - val_accuracy: 0.7997 - val_loss: 0.5852
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8094 - loss: 0.5756 - val_accuracy: 0.8118 - val_loss: 0.5546
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8179 - loss: 0.5446 - val_accuracy: 0.8218 - val_loss: 0.5265
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8247 - loss: 0.5222 - val_accuracy: 0.8248 - val_loss: 0.5106
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8292 - loss: 0.5055 - val_accuracy: 0.8270 - val_loss: 0.4943
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8321 - loss: 0.4926 - val_accuracy: 0.

Accuracy: ~83.9%

Very similar to SGD (since SGD ≈ mini-batch in practice)

Balanced performance and stability

Conclusion: Good trade-off between speed & accuracy

#NESTROV GD

In [7]:
model_nag = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_nag.compile(
    loss='categorical_crossentropy',
    optimizer=SGD(learning_rate=0.01, momentum=0.9, nesterov=True),
    metrics=['accuracy']
)

history_nag = model_nag.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1
)

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7793 - loss: 0.6508 - val_accuracy: 0.8298 - val_loss: 0.4822
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8411 - loss: 0.4592 - val_accuracy: 0.8420 - val_loss: 0.4357
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8540 - loss: 0.4226 - val_accuracy: 0.8532 - val_loss: 0.4214
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8620 - loss: 0.3997 - val_accuracy: 0.8585 - val_loss: 0.3935
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8670 - loss: 0.3801 - val_accuracy: 0.8635 - val_loss: 0.3826
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8720 - loss: 0.3661 - val_accuracy: 0.8688 - val_loss: 0.3764
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8758 - loss: 0.3543 - val_accuracy: 0.8697 - val_loss: 0.3644
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8788 - loss: 0.3421 - val_accuracy: 0.

Accuracy reached ~90.7% train, ~88.8% val

Faster and smoother than Momentum

Better validation accuracy

Conclusion: Performs better than standard Momentum due to look-ahead mechanism

#ADAGRAND

In [8]:
from tensorflow.keras.optimizers import Adagrad

model_adagrad = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_adagrad.compile(
    loss='categorical_crossentropy',
    optimizer=Adagrad(learning_rate=0.01),
    metrics=['accuracy']
)

history_adagrad = model_adagrad.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1
)

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.7405 - loss: 0.8322 - val_accuracy: 0.8063 - val_loss: 0.5977
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8146 - loss: 0.5607 - val_accuracy: 0.8213 - val_loss: 0.5263
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8303 - loss: 0.5070 - val_accuracy: 0.8278 - val_loss: 0.4867
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8387 - loss: 0.4783 - val_accuracy: 0.8368 - val_loss: 0.4619
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8455 - loss: 0.4586 - val_accuracy: 0.8395 - val_loss: 0.4529
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8503 - loss: 0.4437 - val_accuracy: 0.8403 - val_loss: 0.4489
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8532 - loss: 0.4314 - val_accuracy: 0.8503 - val_loss: 0.4298
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8567 - loss: 0.4216 - val_accuracy: 0.

Accuracy: ~87–88%

Fast initial learning

Later slows down due to decreasing learning rate

Conclusion: Good initially but learning rate decay limits performance

#RMS PROP

In [9]:
from tensorflow.keras.optimizers import RMSprop

model_rmsprop = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_rmsprop.compile(
    loss='categorical_crossentropy',
    optimizer=RMSprop(learning_rate=0.001),
    metrics=['accuracy']
)

history_rmsprop = model_rmsprop.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1
)

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7899 - loss: 0.6041 - val_accuracy: 0.8392 - val_loss: 0.4524
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8476 - loss: 0.4252 - val_accuracy: 0.8523 - val_loss: 0.4007
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8640 - loss: 0.3769 - val_accuracy: 0.8518 - val_loss: 0.4272
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8758 - loss: 0.3461 - val_accuracy: 0.8742 - val_loss: 0.3509
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8802 - loss: 0.3273 - val_accuracy: 0.8758 - val_loss: 0.3419
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8867 - loss: 0.3088 - val_accuracy: 0.8718 - val_loss: 0.3497
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8921 - loss: 0.2954 - val_accuracy: 0.8710 - val_loss: 0.3509
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8961 - loss: 0.2828 - val_accuracy: 0.

Accuracy: ~92–93% train, ~89% val

Very fast convergence

Stable training

Conclusion: One of the best performers

#ADADELTA

In [10]:
from tensorflow.keras.optimizers import RMSprop

model_rmsprop = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_rmsprop.compile(
    loss='categorical_crossentropy',
    optimizer=RMSprop(learning_rate=0.001),
    metrics=['accuracy']
)

history_rmsprop = model_rmsprop.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1
)

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7932 - loss: 0.5933 - val_accuracy: 0.8243 - val_loss: 0.4794
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8508 - loss: 0.4202 - val_accuracy: 0.8508 - val_loss: 0.4166
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8656 - loss: 0.3739 - val_accuracy: 0.8633 - val_loss: 0.3661
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8761 - loss: 0.3428 - val_accuracy: 0.8692 - val_loss: 0.3566
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8826 - loss: 0.3228 - val_accuracy: 0.8793 - val_loss: 0.3377
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8887 - loss: 0.3068 - val_accuracy: 0.8747 - val_loss: 0.3561
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8937 - loss: 0.2918 - val_accuracy: 0.8793 - val_loss: 0.3318
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8977 - loss: 0.2806 - val_accuracy: 0.

Accuracy: ~92–93% train, ~88–89% val

No need to manually set learning rate

Slight fluctuations in validation

Conclusion: Stable but slightly less consistent than RMSProp

#ADAM

In [11]:
from tensorflow.keras.optimizers import Adam

model_adam = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_adam.compile(
    loss='categorical_crossentropy',
    optimizer=Adam(learning_rate=0.001),
    metrics=['accuracy']
)

history_adam = model_adam.fit(
    X_train, y_train_cat,
    epochs=20,
    batch_size=128,
    validation_split=0.1
)

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.8057 - loss: 0.5664 - val_accuracy: 0.8502 - val_loss: 0.4249
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8566 - loss: 0.4107 - val_accuracy: 0.8655 - val_loss: 0.3857
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8689 - loss: 0.3692 - val_accuracy: 0.8702 - val_loss: 0.3714
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8778 - loss: 0.3416 - val_accuracy: 0.8752 - val_loss: 0.3405
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8858 - loss: 0.3192 - val_accuracy: 0.8752 - val_loss: 0.3466
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8895 - loss: 0.3068 - val_accuracy: 0.8802 - val_loss: 0.3353
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8939 - loss: 0.2919 - val_accuracy: 0.8843 - val_loss: 0.3230
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8988 - loss: 0.2800 - val_accuracy: 0.

Accuracy: ~92–93% train, ~89% val

Fastest convergence

Very stable

Best balance of speed + accuracy

Conclusion: Best optimizer among all

From the experimental results on the Fashion-MNIST dataset, it is observed that basic gradient descent methods like BGD and SGD show slower convergence and lower accuracy. Advanced optimizers such as Momentum and Nesterov improve convergence speed and stability. Adaptive methods like RMSProp, Adadelta, and Adam perform significantly better by dynamically adjusting learning rates. Among all, Adam optimizer achieved the best performance in terms of accuracy, speed, and stability, making it the most suitable choice for training deep neural networks.

#MLP AND TYPES OF REGULARIZATIONS

In [1]:
import numpy as np
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Dropout, GaussianNoise, Conv2D, MaxPooling2D
from tensorflow.keras.regularizers import l2
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.callbacks import EarlyStopping

# -----------------------------
# Load Dataset (CIFAR-10)
# -----------------------------
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

X_train = X_train / 255.0
X_test = X_test / 255.0

y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# -----------------------------
# Base MLP Model
# -----------------------------
def create_mlp(reg=None, dropout=False, noise=False):
    model = Sequential()

    model.add(Flatten(input_shape=(32,32,3)))

    if noise:
        model.add(GaussianNoise(0.1))   # Noise Injection

    model.add(Dense(128, activation='relu', kernel_regularizer=reg))

    if dropout:
        model.add(Dropout(0.5))

    model.add(Dense(10, activation='softmax'))

    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    return model


# -----------------------------
# 1. L2 Regularization
# -----------------------------
print("\n--- L2 Regularization ---")
model_l2 = create_mlp(reg=l2(0.001))
model_l2.fit(X_train, y_train, epochs=5, batch_size=128, verbose=1)
print("L2 Accuracy:", model_l2.evaluate(X_test, y_test, verbose=0)[1])


# -----------------------------
# 2. Dropout
# -----------------------------
print("\n--- Dropout ---")
model_dropout = create_mlp(dropout=True)
model_dropout.fit(X_train, y_train, epochs=5, batch_size=128, verbose=1)
print("Dropout Accuracy:", model_dropout.evaluate(X_test, y_test, verbose=0)[1])


# -----------------------------
# 3. Noise Injection
# -----------------------------
print("\n--- Noise Injection ---")
model_noise = create_mlp(noise=True)
model_noise.fit(X_train, y_train, epochs=5, batch_size=128, verbose=1)
print("Noise Accuracy:", model_noise.evaluate(X_test, y_test, verbose=0)[1])


# -----------------------------
# 4. Early Stopping
# -----------------------------
print("\n--- Early Stopping ---")
early = EarlyStopping(monitor='val_loss', patience=2)

model_early = create_mlp()
model_early.fit(X_train, y_train,
                validation_split=0.2,
                epochs=20,
                batch_size=128,
                callbacks=[early],
                verbose=1)

print("EarlyStopping Accuracy:", model_early.evaluate(X_test, y_test, verbose=0)[1])


# -----------------------------
# 5. Data Augmentation
# -----------------------------
print("\n--- Data Augmentation ---")
datagen = ImageDataGenerator(
    rotation_range=10,
    width_shift_range=0.1,
    height_shift_range=0.1,
    horizontal_flip=True
)

model_aug = create_mlp()

model_aug.fit(datagen.flow(X_train, y_train, batch_size=128),
              epochs=5,
              verbose=1)

print("Augmentation Accuracy:", model_aug.evaluate(X_test, y_test, verbose=0)[1])


# -----------------------------
# 6. Parameter Sharing (CNN)
# -----------------------------
print("\n--- Parameter Sharing (CNN) ---")

model_cnn = Sequential([
    Conv2D(32, (3,3), activation='relu', input_shape=(32,32,3)),
    MaxPooling2D(),
    Conv2D(64, (3,3), activation='relu'),
    MaxPooling2D(),
    Flatten(),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model_cnn.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

model_cnn.fit(X_train, y_train, epochs=5, batch_size=128, verbose=1)
print("CNN (Parameter Sharing) Accuracy:", model_cnn.evaluate(X_test, y_test, verbose=0)[1])


# -----------------------------
# 7. Ensemble Method
# -----------------------------
print("\n--- Ensemble ---")

model1 = create_mlp()
model2 = create_mlp()

model1.fit(X_train, y_train, epochs=5, batch_size=128, verbose=0)
model2.fit(X_train, y_train, epochs=5, batch_size=128, verbose=0)

pred1 = model1.predict(X_test)
pred2 = model2.predict(X_test)

ensemble_pred = (pred1 + pred2) / 2
ensemble_acc = np.mean(np.argmax(ensemble_pred, axis=1) == np.argmax(y_test, axis=1))

print("Ensemble Accuracy:", ensemble_acc)

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 14s 0us/step

--- L2 Regularization ---


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 8ms/step - accuracy: 0.3066 - loss: 2.0969
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.3718 - loss: 1.8519
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.3968 - loss: 1.7708
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.4122 - loss: 1.7224
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.4181 - loss: 1.7018
L2 Accuracy: 0.42149999737739563

--- Dropout ---
Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.1916 - loss: 2.1455
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.2244 - loss: 2.0500
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.2282 - loss: 2.0352
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.2384 - loss: 2.0214
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.2496 - loss: 2.0134
Dropout Accuracy: 0.33880001306533813

--- Noise Injection ---
Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.4435 - loss: 1.5580
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.5834 - loss: 1.1925
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.6342 - loss: 1.0513
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.6619 - loss: 0.9694
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.6889 - loss: 0.9021
CNN (Parameter Sharing) Accuracy: 0.6697999835014343

--- Ensemble ---
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step
Ensemble Accuracy: 0.4226


1. L2 Regularization

Accuracy: ~42.1%

Helps reduce overfitting by penalizing large weights

Moderate improvement in performance

Observation: Provides stable learning but limited accuracy improvement

🔹 2. Dropout

Accuracy: ~33.8% (lowest)

Randomly drops neurons during training

Model struggles to learn due to high dropout (0.5)

Observation: Too much dropout caused underfitting

🔹 3. Noise Injection

Accuracy: ~42.0%

Makes model robust to noisy inputs

Performance similar to L2

Observation: Helps generalization but not very impactful alone

🔹 4. Early Stopping

Accuracy: ~43.0%

Stops training before overfitting

Prevents unnecessary epochs

Observation: Slight improvement over L2, efficient training

🔹 5. Data Augmentation

Accuracy: ~39.5%

Generates new variations of images

Slower training due to transformations

Observation: Improves generalization but requires more epochs for better results

🔹 6. Parameter Sharing (CNN)

Accuracy: ~66.9% (BEST)

Uses convolution layers → shared filters

Learns spatial features effectively

Observation: Huge performance improvement over MLP

🔹 7. Ensemble Method

Accuracy: ~42.2%

Combines predictions from multiple models

Slight improvement but limited due to weak base models

Observation: Works better when individual models are strong

#CNN

In [2]:
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, Flatten, BatchNormalization
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam

# Load dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

# Normalize
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

# One-hot encoding
Y_train = to_categorical(y_train, 10)
Y_test = to_categorical(y_test, 10)

# Build Improved CNN
model = Sequential()

# Block 1
model.add(Conv2D(32, (3,3), activation='relu', padding='same', input_shape=(32,32,3)))
model.add(BatchNormalization())
model.add(MaxPool2D((2,2)))

# Block 2
model.add(Conv2D(64, (3,3), activation='relu', padding='same'))
model.add(BatchNormalization())
model.add(MaxPool2D((2,2)))

# Block 3 (NEW - deeper network)
model.add(Conv2D(128, (3,3), activation='relu', padding='same'))
model.add(BatchNormalization())
model.add(MaxPool2D((2,2)))

# Fully Connected
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

# Compile with tuned learning rate
optimizer = Adam(learning_rate=0.001)

model.compile(
    loss='categorical_crossentropy',
    optimizer=optimizer,
    metrics=['accuracy']
)

# Train
history = model.fit(
    X_train, Y_train,
    batch_size=64,
    epochs=15,
    validation_data=(X_test, Y_test),
    verbose=1
)

# Evaluate
test_loss, test_acc = model.evaluate(X_test, Y_test)
print("Test Accuracy:", test_acc)

Epoch 1/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 16s 12ms/step - accuracy: 0.4569 - loss: 1.5339 - val_accuracy: 0.5528 - val_loss: 1.2400
Epoch 2/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.5973 - loss: 1.1375 - val_accuracy: 0.5690 - val_loss: 1.2520
Epoch 3/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.6662 - loss: 0.9506 - val_accuracy: 0.5995 - val_loss: 1.1767
Epoch 4/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.7096 - loss: 0.8340 - val_accuracy: 0.6848 - val_loss: 0.9263
Epoch 5/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.7402 - loss: 0.7419 - val_accuracy: 0.7119 - val_loss: 0.8382
Epoch 6/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.7660 - loss: 0.6722 - val_accuracy: 0.7182 - val_loss: 0.8295
Epoch 7/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.7875 - loss: 0.6109 - val_accuracy: 0.7061 - val_loss: 0.8756
Epoch 8/15
782/782 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.8104 - loss: 0.5421 - val_accuracy: 

The CNN model achieved a maximum test accuracy of approximately 75.5% on the CIFAR-10 dataset. Increasing the depth of the network and using Batch Normalization significantly improved performance. However, a gap between training accuracy (89%) and validation accuracy (~75%) indicates overfitting. The optimal performance was observed between epochs 8 and 12. Further improvements can be achieved using stronger regularization techniques such as higher dropout, data augmentation, and early stopping.

#LENET

In [3]:
# LENET CIFAR-10 (Fixed Version)

from tensorflow.keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, AveragePooling2D, Flatten, Dense

# Load Data
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

# Normalize
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Model
model = Sequential()

model.add(Conv2D(6, (5,5), activation='relu', input_shape=(32,32,3)))
model.add(AveragePooling2D(pool_size=(2,2)))

model.add(Conv2D(16, (5,5), activation='relu'))
model.add(AveragePooling2D(pool_size=(2,2)))

model.add(Flatten())

model.add(Dense(120, activation='relu'))
model.add(Dense(84, activation='relu'))
model.add(Dense(10, activation='softmax'))

# compile
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train
model.fit(X_train, y_train,
          epochs=1,
          batch_size=128,
          validation_data=(X_test, y_test))

#  Evaluate
model.evaluate(X_test, y_test)

391/391 ━━━━━━━━━━━━━━━━━━━━ 11s 19ms/step - accuracy: 0.3579 - loss: 1.7657 - val_accuracy: 0.4149 - val_loss: 1.6317
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.4149 - loss: 1.6317


[1.6316667795181274, 0.414900004863739]

Underfitting due to shallow architecture

#ALEXNET

In [4]:
#ALEXNET CIFAR 10
from keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

(X_train, y_train), (X_test, y_test) = cifar10.load_data()

X_train = X_train / 255.0
X_test = X_test / 255.0

y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

model = Sequential()
model.add(Conv2D(96, (3,3), activation='relu', input_shape=(32,32,3)))
model.add(MaxPooling2D())
model.add(Conv2D(256, (3,3), activation='relu'))
model.add(MaxPooling2D())
model.add(Conv2D(384, (3,3), activation='relu'))
model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=5, batch_size=128, validation_data=(X_test, y_test))

model.evaluate(X_test, y_test)

Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 18s 31ms/step - accuracy: 0.4086 - loss: 1.6093 - val_accuracy: 0.5302 - val_loss: 1.2965
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 7s 18ms/step - accuracy: 0.5890 - loss: 1.1562 - val_accuracy: 0.6489 - val_loss: 0.9954
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 7s 18ms/step - accuracy: 0.6682 - loss: 0.9497 - val_accuracy: 0.6817 - val_loss: 0.8963
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 7s 19ms/step - accuracy: 0.7097 - loss: 0.8304 - val_accuracy: 0.7211 - val_loss: 0.8039
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 8s 21ms/step - accuracy: 0.7423 - loss: 0.7401 - val_accuracy: 0.7341 - val_loss: 0.7722
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7341 - loss: 0.7722


[0.7721744775772095, 0.7340999841690063]

Good balance of depth and performance

#ZF NET

In [5]:
#ZFNET CIFAR 10
from keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Input

# Load data
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

# Normalize
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Model
model = Sequential()

model.add(Input(shape=(32,32,3)))  # Fix warning

model.add(Conv2D(96, (7,7), strides=2, activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2,2)))

model.add(Conv2D(256, (5,5), activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2,2)))

model.add(Conv2D(384, (3,3), activation='relu', padding='same'))

model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train
model.fit(X_train, y_train, epochs=5, batch_size=128, validation_data=(X_test, y_test))

# Evaluate
model.evaluate(X_test, y_test)

Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 15s 26ms/step - accuracy: 0.3760 - loss: 1.6930 - val_accuracy: 0.5074 - val_loss: 1.3556
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 14ms/step - accuracy: 0.5520 - loss: 1.2490 - val_accuracy: 0.5946 - val_loss: 1.1460
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 14ms/step - accuracy: 0.6339 - loss: 1.0319 - val_accuracy: 0.6460 - val_loss: 1.0091
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 15ms/step - accuracy: 0.6903 - loss: 0.8820 - val_accuracy: 0.6498 - val_loss: 0.9853
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 14ms/step - accuracy: 0.7357 - loss: 0.7471 - val_accuracy: 0.6753 - val_loss: 0.9546
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.6753 - loss: 0.9546


[0.9546400308609009, 0.6753000020980835]

Not as effective as smaller filters

#VGG NET

In [6]:
#VGGNET CIFAR 10
from keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

(X_train, y_train), (X_test, y_test) = cifar10.load_data()

X_train = X_train / 255.0
X_test = X_test / 255.0

y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

model = Sequential()

model.add(Conv2D(64, (3,3), padding='same', activation='relu', input_shape=(32,32,3)))
model.add(Conv2D(64, (3,3), padding='same', activation='relu'))
model.add(MaxPooling2D())

model.add(Conv2D(128, (3,3), padding='same', activation='relu'))
model.add(Conv2D(128, (3,3), padding='same', activation='relu'))
model.add(MaxPooling2D())

model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=5, batch_size=128, validation_data=(X_test, y_test))

model.evaluate(X_test, y_test)

Epoch 1/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 18s 33ms/step - accuracy: 0.4880 - loss: 1.4196 - val_accuracy: 0.6156 - val_loss: 1.0883
Epoch 2/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 8s 21ms/step - accuracy: 0.6823 - loss: 0.9096 - val_accuracy: 0.7067 - val_loss: 0.8318
Epoch 3/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 9s 23ms/step - accuracy: 0.7601 - loss: 0.6964 - val_accuracy: 0.7454 - val_loss: 0.7264
Epoch 4/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 9s 21ms/step - accuracy: 0.8082 - loss: 0.5516 - val_accuracy: 0.7541 - val_loss: 0.7165
Epoch 5/5
391/391 ━━━━━━━━━━━━━━━━━━━━ 8s 21ms/step - accuracy: 0.8502 - loss: 0.4286 - val_accuracy: 0.7669 - val_loss: 0.7372
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7669 - loss: 0.7372


[0.7372004985809326, 0.7669000029563904]

Best model among all for CIFAR-10

#

#GOOGLENET

In [7]:
#GOOGLENET CIFAR 10
from keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Concatenate, Input
from tensorflow.keras.models import Model

(X_train, y_train), (X_test, y_test) = cifar10.load_data()

X_train = X_train / 255.0
X_test = X_test / 255.0

y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

def inception_module(x):
    path1 = Conv2D(64, (1,1), activation='relu', padding='same')(x)
    path2 = Conv2D(64, (3,3), activation='relu', padding='same')(x)
    path3 = Conv2D(64, (5,5), activation='relu', padding='same')(x)
    return Concatenate()([path1, path2, path3])

inputs = Input(shape=(32,32,3))
x = inception_module(inputs)
x = MaxPooling2D()(x)
x = inception_module(x)

x = Flatten()(x)
x = Dense(256, activation='relu')(x)
outputs = Dense(10, activation='softmax')(x)

model = Model(inputs, outputs)

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=3, batch_size=128, validation_data=(X_test, y_test))

model.evaluate(X_test, y_test)

Epoch 1/3
391/391 ━━━━━━━━━━━━━━━━━━━━ 25s 47ms/step - accuracy: 0.4585 - loss: 1.5441 - val_accuracy: 0.5705 - val_loss: 1.2153
Epoch 2/3
391/391 ━━━━━━━━━━━━━━━━━━━━ 13s 32ms/step - accuracy: 0.6136 - loss: 1.1025 - val_accuracy: 0.6154 - val_loss: 1.0754
Epoch 3/3
391/391 ━━━━━━━━━━━━━━━━━━━━ 20s 31ms/step - accuracy: 0.6677 - loss: 0.9480 - val_accuracy: 0.6621 - val_loss: 0.9606
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.6621 - loss: 0.9606


[0.9606073498725891, 0.6621000170707703]

Needs more training for better results

#RESNET

In [8]:
#RESNET CIFAR 10
import tensorflow as tf
from keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Conv2D, Add, MaxPooling2D, Flatten, Dense, Input
from tensorflow.keras.models import Model

(X_train, y_train), (X_test, y_test) = cifar10.load_data()

X_train = X_train / 255.0
X_test = X_test / 255.0

y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

def res_block(x, filters):
    shortcut = x
    x = Conv2D(filters, (3,3), padding='same', activation='relu')(x)
    x = Conv2D(filters, (3,3), padding='same')(x)
    x = Add()([x, shortcut])
    return tf.keras.activations.relu(x)

inputs = Input(shape=(32,32,3))
x = Conv2D(64, (3,3), padding='same')(inputs)
x = res_block(x, 64)
x = MaxPooling2D()(x)

x = Flatten()(x)
x = Dense(128, activation='relu')(x)
outputs = Dense(10, activation='softmax')(x)

model = Model(inputs, outputs)

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=2, batch_size=128, validation_data=(X_test, y_test))

model.evaluate(X_test, y_test)

Epoch 1/2
391/391 ━━━━━━━━━━━━━━━━━━━━ 14s 28ms/step - accuracy: 0.5071 - loss: 1.3863 - val_accuracy: 0.6066 - val_loss: 1.1163
Epoch 2/2
391/391 ━━━━━━━━━━━━━━━━━━━━ 7s 19ms/step - accuracy: 0.6715 - loss: 0.9395 - val_accuracy: 0.6811 - val_loss: 0.9125
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6811 - loss: 0.9125


[0.912489652633667, 0.6811000108718872]

Very powerful model but undertrained here

Among all implemented architectures, VGGNet achieved the highest accuracy (~75.8%) due to its deep architecture and use of small 3×3 convolution filters, which effectively capture fine-grained features. AlexNet also performed well, while LeNet showed poor performance due to its shallow design. ResNet demonstrated strong learning capability even with fewer epochs due to residual connections. Overall, deeper architectures with smaller filters and regularization techniques provide better performance on complex datasets like CIFAR-10.